go-controller/pkg/libovsdbops/loadbalancer.go

```diff
@@ -2,6 +2,7 @@ package libovsdbops
 
```

Comment:
> let's hope that the move to IC can save us from having to do the former as well. but woohoo! happy to see the scale gains we'd get from this!
> 

Ack.

> RE: commit message: `- before this change we would configure N * S load balancers (one per node for every service)` wasn't it `S * 2N -> one per switch/router`? I used to remember 1 NP generating 7LBs on a 3 node cluster (1 clusterwide LB and 6 nodeport LBs)

You're right.

> 
> RE: Templates (nit: include this awesome info on the commit message as well) `Instead of generating a load balancer per service per node we now generate at most four (template) load balancers (IPv4/IPv6 applicable to gateway routers or applicable to node switches). Each template load balancer uses as VIP a template variable denoting the node IP, e.g., NODEIP_IPv4 and as backend another template variable, e.g., SVC_node_port_svc__UDP__30957__node__router__template__IPv4. These template variables have potentially different values on different chassis. For example, NODEIP_IPv4 will expand to chassis-X's IP on chassis-X and to chassis-Y's IP on chassis-Y. Similarly, SVC_node_port_svc__UDP__30957__node__router__template__IPv4 will expand on chassis-X to the set of backends that are required by the NodePort service on chassis-X and that might be different than the set of backends required on chassis-Y.`
> 

Done.

> Does the number of templates cause any scale issues? -> per commit message I guess its insignificant. I am thinking let's say we start supporting nodePorts for all reachable IPs of a node, that means its a new template per nodeIP right? cc @zeeke : whenever you decide to rebase #3328 you might wanna rebase that on top of these changes.
> 

It should not cause scale issues.  I added a link to the original OVN patch cover letter and ovscon presentation:
https://mail.openvswitch.org/pipermail/ovs-dev/2022-November/399522.html
https://www.openvswitch.org/support/ovscon2022/slides/Horizontal-scaling.pdf

> I guess q is, as long as either the VIP or IP is different we'd need a new template for the service right? So 4 template LBs per service (dualstack; 2LBs if single stack) if there is only support for 1 nodeIP vip like today. If its X nodeIPs is it 4X template LBs per service or is it still 4LBs per service since only the entry within the vip will change? Will this scale with number of nodeIPs to something like this:
> 
> `vips : {"^NODEIP1_IPv4:32759"="^Service__default_example_service_hello_world__TCP__32759__node__switch__template__IPv4", "^NODEIP2_IPv4:32759"="^Service__default_example_service_hello_world__TCP__32759__node__switch__template__IPv4"}`

Yes, it will require a new template variable and LB per node IP.  So you'll need 4X template LBs if X is the max number of IPs a node can have.  However, I wonder how many IPs a node can have.. will it ever be more than "a few"?